### PR TITLE
ACE Laser & Helfire

### PR DESCRIPTION
**Pull request by mattysmith22**

### CODE REVIEW

File: @AH-64D Apache Longbow/Addons/fza_ah64_controls/scripting/functions/hellfire/fn_hellfireARHSeeker.sqf [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+_args params ["_firedEH", "_launchParams", "", "_seekerParams", "_stateParams"];
+_firedEH params ["_shooter","","","","","","_projectile"];
+_seekerParams params ["_seekerAngle", "", "_seekerMaxRange"];
+_seekerStateParams params ["_TargetObj", "_TargetPos", "_TargetType","_LaunchLobl"];
+
+
+private _heli 				= vehicle _shooter;
+private _returnTargetPos  	= _TargetPos;
+private _SelectedTarget		= objNull;
+private _LOSCounterVAR		= "fza_ah64_" + str _projectile + "_Los_Lost_FrameCounter";
```
[the remark this answers is not in the document]
Comment (mattysmith22) [reply @AH-64D Apache Longbow/Addons/fza_ah64_controls/scripting/functions/hellfire/fn_hellfireARHSeeker.sqf:29]: Correction. For anything requiring state, there exists within ACE missile guidance `_seekerStateParams` to store it in. Please could you add that there?

File: @AH-64D Apache Longbow/Addons/fza_ah64_controls/scripting/functions/hellfire/fn_hellfireARHSeeker.sqf [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+		private _newScanTargets  = nearestObjects [_TargetPos, ["land","air","ship"], 500];
+        private _validTargets 	 = _newScanTargets apply {
+            if ((([_projectile, [getpos _x, "", speed _x, _x], true] call fza_fnc_hellfireLimaLoblCheck) # 1) == true) then {
+                _x
+            };
+        };
+
+		private _Primarytargets = _validTargets select {
+			private _targTypeCompair = _x call BIS_fnc_objectType;
+			(_targetType isEqualTo _targTypeCompair)
```
Comment (mattysmith22) [inline @AH-64D Apache Longbow/Addons/fza_ah64_controls/scripting/functions/hellfire/fn_hellfireARHSeeker.sqf:56]: please can you change this to `_targTypeCompare`

File: @AH-64D Apache Longbow/Addons/fza_ah64_controls/scripting/functions/hellfire/fn_hellfireARHSeeker.sqf [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+		};
+		private _secondarytargets = _validTargets select {
+			private _targTypeCompair = _x call BIS_fnc_objectType;
+			!(_targetType isEqualTo _targTypeCompair)
+		};
+		If (_Primarytargets isNotEqualTo []) then {
```
Comment (mattysmith22) [inline @AH-64D Apache Longbow/Addons/fza_ah64_controls/scripting/functions/hellfire/fn_hellfireARHSeeker.sqf:62]: `if`

File: @AH-64D Apache Longbow/Addons/fza_ah64_controls/scripting/functions/hellfire/fn_hellfireARHSeeker.sqf [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+				_SelectedTarget = [_secondarytargets, _TargetPos] call BIS_fnc_nearestPosition;
+			};
+		};
+    };
+};
+
+if !(isNull _SelectedTarget) then {
+	//private _aimPosTarget = aimpos _SelectedTarget;
+	private _centerOfObject = getCenterOfMass _SelectedTarget;
+    private _aimPosTarget = _SelectedTarget modelToWorldWorld _centerOfObject;
```
Comment (mattysmith22) [inline @AH-64D Apache Longbow/Addons/fza_ah64_controls/scripting/functions/hellfire/fn_hellfireARHSeeker.sqf:75]: The indentation on this seems to be incorrect in the viewer

File: @AH-64D Apache Longbow/Addons/fza_ah64_controls/scripting/functions/hellfire/fn_hellfireCheckSeekerAngle.sqf [hunk 5]
```diff
@@ -0,0 +1,29 @@
+/* ----------------------------------------------------------------------------
+Function: fza_fnc_hellfireCheckSeekerAngle
+
+Description:
+    returns true,false if target is in seeker angle of the object
+    Derived from the ace check seeker angle but to use model vectoring instead of flight vector
+
+Parameters:
```
Comment (mattysmith22) [inline @AH-64D Apache Longbow/Addons/fza_ah64_controls/scripting/functions/hellfire/fn_hellfireCheckSeekerAngle.sqf:8]: Please can you fill this in

File: @AH-64D Apache Longbow/Addons/fza_ah64_controls/scripting/functions/hellfire/fn_hellfireGetAttackProfile.sqf [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+Examples:
+    [_heli] call fza_fnc_HellfireGetAttackProfile
+
+Author:
+	Rosd6(Dryden)
+---------------------------------------------------------------------------- */
+params ["_heli"];
+
+private _trajectory = _heli getvariable "fza_ah64_hellfireTrajectory";
+Private _result = false;
```
Comment (mattysmith22) [inline @AH-64D Apache Longbow/Addons/fza_ah64_controls/scripting/functions/hellfire/fn_hellfireGetAttackProfile.sqf:22]: `private`

File: @AH-64D Apache Longbow/Addons/fza_ah64_controls/scripting/functions/hellfire/fn_hellfireGetAttackProfile.sqf [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+Function: fza_fnc_HellfireGetAttackProfile
+
+Description:
+	grabs the ace name for the current attack mode.
+
+Parameters:
+    _heli - the helicopter
+
+Returns:
+	Nothing
```
Comment (mattysmith22) [inline @AH-64D Apache Longbow/Addons/fza_ah64_controls/scripting/functions/hellfire/fn_hellfireGetAttackProfile.sqf:11]: Please can you fill this in

File: @AH-64D Apache Longbow/Addons/fza_ah64_controls/scripting/functions/hellfire/fn_hellfireLimaLoblCheck.sqf [hunk 8]
```diff
@@ -0,0 +1,51 @@
+/* ----------------------------------------------------------------------------
+Function: fza_fnc_hellfireLimaLoblCheck
+
+Description:
+    Defines what the radar can see. Can either be given a
```
Comment (mattysmith22) [inline @AH-64D Apache Longbow/Addons/fza_ah64_controls/scripting/functions/hellfire/fn_hellfireLimaLoblCheck.sqf:5]: Please can you fill this documentation in

File: @AH-64D Apache Longbow/Addons/fza_ah64_controls/scripting/functions/hellfire/fn_hellfireSALFired.sqf [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+private _attackProfile = [_heli] call fza_fnc_HellfireGetAttackProfile;
+
+if (_missilelobl == true) then {
+	_attackProfile = "hellfire";
+};
+
+_launchParams set [3, _attackProfile];
+
+//Grab and set laser code for hellfire
+_index = _heli getvariable "fza_ah64_laserMissilePrimaryCode";
```
Comment (mattysmith22) [inline @AH-64D Apache Longbow/Addons/fza_ah64_controls/scripting/functions/hellfire/fn_hellfireSALFired.sqf:36]: `getVariable`

File: @AH-64D Apache Longbow/Addons/fza_ah64_controls/scripting/functions/hellfire/fn_hellfireSALFired.sqf [hunk 10]
```diff
@@ -0,0 +1,47 @@
+/* ----------------------------------------------------------------------------
+Function: fza_fnc_hellfireSALFired
+
+Description:
+    The handling of Ace Laser Seeker for our Systems
```
Comment (mattysmith22) [inline @AH-64D Apache Longbow/Addons/fza_ah64_controls/scripting/functions/hellfire/fn_hellfireSALFired.sqf:5]: "Initialises the seeker of the ACE SAL hellfire for the apache`

File: @AH-64D Apache Longbow/Addons/fza_ah64_controls/scripting/functions/hellfire/fn_hellfireSALPrelaunch.sqf [hunk 11]
```diff
@@ -0,0 +1,33 @@
+/* ----------------------------------------------------------------------------
+Function: fza_fnc_hellfireSALPrelaunch
+
+Description:
+    Defines what the SAL missile can see.
```
Comment (mattysmith22) [inline @AH-64D Apache Longbow/Addons/fza_ah64_controls/scripting/functions/hellfire/fn_hellfireSALPrelaunch.sqf:5]: This seems to conflict with the name of the function.

Could we change it to

```
Function: fza_fnc_hellfireSALShouldStartLobl
Description:
    Defines whether a SAL missile should start in LOBL
    (can see valid laser energy)
```

File: @AH-64D Apache Longbow/Addons/fza_ah64_controls/scripting/functions/ihadss/fn_ihadssDraw.sqf [hunk 12]
```diff
@@ -514,65 +514,39 @@ if (_was == WAS_WEAPON_MSL) then {
     };
     _targpos = _scPos;
 
-    if (isNull _mistargPos) then {
+    if (isNull _missileTarget || (_heli getVariable "fza_ah64_selectedMissile" == "fza_agm114l_wep" && isVehicleRadarOn _heli == false)) then {
```
Comment (mattysmith22) [inline @AH-64D Apache Longbow/Addons/fza_ah64_controls/scripting/functions/ihadss/fn_ihadssDraw.sqf:517]: What is this change here achieving?

File: @AH-64D Apache Longbow/Addons/fza_ah64_controls/scripting/functions/ihadss/fn_ihadssDraw.sqf [hunk 13]
```diff
@@ -514,65 +514,39 @@ if (_was == WAS_WEAPON_MSL) then {
...
         ((uiNameSpace getVariable "fza_ah64_raddisp") displayCtrl 131) ctrlSetText "";
     } else {
-        _terrainobscure = terrainIntersectasl[[(getPosASL _heli select 0) + ((sin getdir _heli) * 6), (getPosASL _heli select 1) + ((cos getdir _heli) * 6), (getPosASL _heli select 2)], [(getPosASL _mistargPos select 0), (getPosASL _mistargPos select 1), (getPosASL _mistargPos select 2) + 1]];
-        _obscureobjs = lineIntersectsWith[[(getPosASL _heli select 0) + ((sin getdir _heli) * 6), (getPosASL _heli select 1) + ((cos getdir _heli) * 6), (getPosASL _heli select 2)], getPosASL _mistargPos, _heli, _mistargPos];
-        _distOffAxis = abs ([[_heli, getPos _heli # 0, getPos _heli # 1, getPos _mistargPos # 0, getPos _mistargPos # 1] call fza_fnc_relativeDirection] call CBA_fnc_simplifyAngle180);
-        _targetDist = _heli distance getPos _mistargPos;
-	    _targ = _heli getVariable "fza_ah64_fcrNts";
+        _targetDist = _heli distance getPos _missileTarget;
+        _LoblCheckLima = [_heli, [getpos _missileTarget, "", speed _missileTarget, _missileTarget]] call fza_fnc_hellfireLimaLoblCheck;
```
Comment (mattysmith22) [inline @AH-64D Apache Longbow/Addons/fza_ah64_controls/scripting/functions/ihadss/fn_ihadssDraw.sqf:521]: please can you standardise this variable name to `_loblCheckLima`

File: @AH-64D Apache Longbow/Addons/fza_ah64_controls/scripting/functions/ihadss/fn_ihadssDraw.sqf [hunk 14]
```diff
@@ -514,65 +514,39 @@ if (_was == WAS_WEAPON_MSL) then {
...
+        _LoblCheck = [_heli] call fza_fnc_hellfireSalPrelaunch;
```
Comment (mattysmith22) [inline @AH-64D Apache Longbow/Addons/fza_ah64_controls/scripting/functions/ihadss/fn_ihadssDraw.sqf:522]: please can you standardise this variable name to `_loblCheck `

Comment (mattysmith22) [inline @AH-64D Apache Longbow/Addons/fza_ah64_controls/scripting/functions/ihadss/fn_ihadssDraw.sqf:522]: Why are we always computing the checks for both? Can we do these inside the stages later for efficiency purposes?

File: @AH-64D Apache Longbow/Addons/fza_ah64_controls/scripting/functions/ihadss/fn_ihadssDraw.sqf [hunk 15]
```diff
@@ -514,65 +514,39 @@ if (_was == WAS_WEAPON_MSL) then {
...
+        _distOffAxis = abs (_heli getRelDir _missileTarget call CBA_fnc_simplifyAngle180);
+        _LoalLimitOfset = 7.5;
```
Comment (mattysmith22) [inline @AH-64D Apache Longbow/Addons/fza_ah64_controls/scripting/functions/ihadss/fn_ihadssDraw.sqf:524]: Please can you standardise this to `_loalLimitOffset`

File: @AH-64D Apache Longbow/Addons/fza_ah64_controls/scripting/functions/weapon/fn_weaponTrajectoryChange.sqf [hunk 16]
```diff
@@ -18,11 +18,6 @@ Author:
...
-private _nextFireMode = switch (_heli getVariable "fza_ah64_hellfireTrajectory") do {
-	case "HI": {"HI"};
-	case "LO": {"LO"};
-	case "DIR": {"DIR"};
-	default {["Unknown missile fire mode: %1", _fireMode] call BIS_fnc_error};
-};
-_heli setVariable ["fza_ah64_hellfireTrajectory", _nextFireMode, true];
-[_heli] call fza_fnc_weaponUpdateSelected;
\ No newline at end of file
+_nextFireMode = [["LO","DIR","HI"], _heli getVariable "fza_ah64_hellfireTrajectory"] call fza_fnc_cycle;
```
Comment (mattysmith22) [inline @AH-64D Apache Longbow/Addons/fza_ah64_controls/scripting/functions/weapon/fn_weaponTrajectoryChange.sqf:21]: This needs to be defined as `private`

File: @AH-64D Apache Longbow/Addons/fza_ah64_controls/scripting/functions/weapon/fn_weaponTurretAim.sqf [hunk 17]
```diff
@@ -27,7 +27,7 @@ private _inhibit = "";
 private _weaponsProcessorFailed = _heli getVariable "fza_ah64_rwp_fail" && _heli getVariable "fza_ah64_lwp_fail";
 
 private _usingRocket = currentweapon _heli isKindOf["fza_hydra70", configFile >> "CfgWeapons"] || currentWeapon _heli == "fza_rkt_safe";
-private _usingCannon = currentweapon _heli in ["fza_m230", "fza_burstlimiter", "fza_gun_safe"];
+private _usingCannon = currentweapon _heli in ["fza_m230", "fza_burstlimiter"];
```
Comment (mattysmith22) [inline @AH-64D Apache Longbow/Addons/fza_ah64_controls/scripting/functions/weapon/fn_weaponTurretAim.sqf:30]: If we are no longer having different safe modes, we need to change how we decide whether we are `_usingCannon` or not.

This introduces a bug where I can WAS gun, be in safe mode, set sight to FXD and not have the "GUN FIXED" warning appear

Comment (mattysmith22) [reply @AH-64D Apache Longbow/Addons/fza_ah64_controls/scripting/functions/weapon/fn_weaponTurretAim.sqf:30]: If the removal of the safe modes were not absolutely required for ace hellfire, I would recomment to not remove the safe modes in this PR, then we can worry about this one in the future

**Review by mattysmith22 — COMMENTED** [review]
I am unable to request changes as I opened the PR. So far so good on the guidance.

* I have a couple of things that i have mentioned should be made into separate PRs
* There are a lot of issues with things being out of standard for our code, namely
    * Variable name capitalisation
    * Documentation of functions
    * Styling (including indentation)

I may look into making a checklist for PRs to help with identifying and fixing these common issues if that would be useful?

File: @AH-64D Apache Longbow/Addons/fza_ah64_controls/scripting/functions/ihadss/fn_ihadssDraw.sqf (hunk 12 above)
Comment (Ollieollieolllie) [reply @AH-64D Apache Longbow/Addons/fza_ah64_controls/scripting/functions/ihadss/fn_ihadssDraw.sqf:517]: the prevention of the Lock box if lima is selected and the radar is off, that being said i think i added this in error

**Review by Ollieollieolllie — APPROVED** [review]
everything looks good ;)

**Pull request merged**